File: Captcha/Abstraction/AbstractCaptcha.cs

namespace Captcha
{


    public abstract class AbstractCaptcha
    {

        public string m_text;
        private System.Drawing.Imaging.ImageFormat m_imageFormat;


        public virtual string Text
        {
            get
            {
                if (string.IsNullOrEmpty(this.m_text))
                    this.m_text = this.RandomToken;

                return this.m_text;
            }
            set { this.m_text = value; }
        }


        public virtual string RandomToken
        {
            get
            {
                System.Random r = new System.Random();
                string s = "";
                for (int j = 0; j < 5; j++)
                {
                    int i = r.Next(3);
                    int ch;
                    switch (i)
                    {
                        case 1:
                            ch = r.Next(0, 9);
                            s = s + ch.ToString();
                            break;
                        case 2:
                            ch = r.Next(65, 90);
                            s = s + System.Convert.ToChar(ch).ToString();
                            break;
                        case 3:
                            ch = r.Next(97, 122);
                            s = s + System.Convert.ToChar(ch).ToString();
                            break;
                        default:
                            ch = r.Next(97, 122);
                            s = s + System.Convert.ToChar(ch).ToString();
                            break;
                    }
                    r.NextDouble();
                    r.Next(100, 1999);
                }
                return s;
            }
        }


        public virtual System.Drawing.Imaging.ImageFormat Format
        {
            get
            {
                if (m_imageFormat == null)
                    m_imageFormat = System.Drawing.Imaging.ImageFormat.Png;

                return m_imageFormat;
            }
            set
            {
                this.m_imageFormat = value;
            }
        }


        private string m_MimeType;

        public virtual string MimeType
        {
            get
            {
                if (!string.IsNullOrEmpty(m_MimeType))
                {
                    return this.m_MimeType;
                }

                System.Guid formatUID = this.Format.Guid;

                System.Drawing.Imaging.ImageCodecInfo[] codecs = System.Drawing.Imaging.ImageCodecInfo.GetImageEncoders();
                for (int i = 0; i < codecs.Length; ++i)
                {
                    if (codecs[i].FormatID == formatUID)
                    {
                        this.m_MimeType = codecs[i].MimeType;
                        return this.m_MimeType;
                    }

                }

                this.m_MimeType = "image/unknown";
                return this.m_MimeType;
            }
        }


        public abstract System.Drawing.Image Image
        {
            get;
            set;
        }



        public virtual byte[] ImageBytes
        {
            get
            {
                byte[] imageBytes = null;

                using (System.IO.MemoryStream ms = new System.IO.MemoryStream())
                {
                    this.Image.Save(ms, this.Format);
                    imageBytes = ms.ToArray();
                }

                return imageBytes;
            }
        }

        public virtual string Base64()
        {
            return System.Convert.ToBase64String(this.ImageBytes);
        }


    }


}

File: vbCaptcha/Captcha/AbstractCaptcha.vb

Public MustInherit Class AbstractCaptcha

    Private m_guid As System.Nullable(Of System.Guid)
    Private m_text As String
    Private m_imageFormat As System.Drawing.Imaging.ImageFormat
    Private m_MimeType As String


    Public Overridable Property UID() As System.Guid
        Get
            If Me.m_guid.HasValue Then
                Return Me.m_guid.Value
            End If

            Me.m_guid = System.Guid.NewGuid()
            Return Me.m_guid.Value
        End Get
        Set(value As System.Guid)
            Me.m_guid = value
        End Set
    End Property


    Public Overridable Property Text() As String
        Get
            If String.IsNullOrEmpty(Me.m_text) Then
                Me.m_text = Me.RandomToken
            End If

            Return Me.m_text
        End Get
        Set(value As String)
            Me.m_text = value
        End Set
    End Property


    Public Overridable ReadOnly Property RandomToken() As String
        Get
            Dim r As New System.Random()
            Dim s As String = ""
            For j As Integer = 0 To 4
                Dim i As Integer = r.[Next](3)
                Dim ch As Integer
                Select Case i
                    Case 1
                        ch = r.[Next](0, 9)
                        s = s & ch.ToString()
                        Exit Select
                    Case 2
                        ch = r.[Next](65, 90)
                        s = s & System.Convert.ToChar(ch).ToString()
                        Exit Select
                    Case 3
                        ch = r.[Next](97, 122)
                        s = s & System.Convert.ToChar(ch).ToString()
                        Exit Select
                    Case Else
                        ch = r.[Next](97, 122)
                        s = s & System.Convert.ToChar(ch).ToString()
                        Exit Select
                End Select
                r.NextDouble()
                r.[Next](100, 1999)
            Next
            Return s
        End Get
    End Property ' RandomToken 

    Public Overridable Property Format() As System.Drawing.Imaging.ImageFormat
        Get
            If m_imageFormat Is Nothing Then
                m_imageFormat = System.Drawing.Imaging.ImageFormat.Png
            End If

            Return m_imageFormat
        End Get
        Set(value As System.Drawing.Imaging.ImageFormat)
            Me.m_imageFormat = value
        End Set
    End Property


    Public Overridable ReadOnly Property MimeType() As String
        Get
            If Not String.IsNullOrEmpty(m_MimeType) Then
                Return Me.m_MimeType
            End If

            Dim formatUID As System.Guid = Me.Format.Guid

            Dim codecs As System.Drawing.Imaging.ImageCodecInfo() = System.Drawing.Imaging.ImageCodecInfo.GetImageEncoders()
            For i As Integer = 0 To codecs.Length - 1
                If codecs(i).FormatID = formatUID Then
                    Me.m_MimeType = codecs(i).MimeType
                    Return Me.m_MimeType

                End If
            Next

            Me.m_MimeType = "image/unknown"
            Return Me.m_MimeType
        End Get
    End Property


    Public MustOverride Property Image() As System.Drawing.Image


    Public Overridable ReadOnly Property ImageBytes() As Byte()
        Get
            Dim imageBytes__1 As Byte() = Nothing

            Using ms As New System.IO.MemoryStream()
                Me.Image.Save(ms, Me.Format)
                imageBytes__1 = ms.ToArray()
            End Using

            Return imageBytes__1
        End Get
    End Property


    Public Overridable ReadOnly Property Base64() As String
        Get
            Return System.Convert.ToBase64String(Me.ImageBytes)
        End Get
    End Property


End Class ' AbstractCaptcha 
